e2e/process_utils_test.go

```diff
@@ -0,0 +1,224 @@
+	endpoint := fmt.Sprintf("http://%s/version", g.ClientAddress)
+	resp, err := http.Get(endpoint)
+	if err != nil {
+		return ""
+	}
+	defer resp.Body.Close()
+
+	g.uuid = resp.Header.Get("X-Gluster-Peer-Id")
+	return g.uuid
+}
+
+func (g *gdProcess) IsRestServerUp() bool {
+
+	endpoint := fmt.Sprintf("http://%s/v1/peers", g.ClientAddress)
+	resp, err := http.Get(endpoint)
+	if err != nil {
+		return false
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode/100 == 5 {
+		return false
+	}
+
+	return true
+}
+
+func spawnGlusterd(configFilePath string, cleanStart bool) (*gdProcess, error) {
+
+	fContent, err := ioutil.ReadFile(configFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	g := gdProcess{}
+	if err = toml.Unmarshal(fContent, &g); err != nil {
+		return nil, err
+	}
+
+	// The config files in e2e/config contain relative paths, convert them
+	// to absolute paths.
+	g.updateDirs()
+
+	if cleanStart {
+		g.EraseLocalStateDir() // cleanup leftovers from previous test
+	}
+
+	if err := os.MkdirAll(path.Join(g.LocalStateDir, "log"), os.ModeDir|os.ModePerm); err != nil {
+		return nil, err
+	}
+
+	absConfigFilePath, err := filepath.Abs(configFilePath)
+	if err != nil {
+		return nil, err
+	}
+	args := []string{
+		"--config", absConfigFilePath,
+		"--localstatedir", g.LocalStateDir,
+		"--rundir", g.Rundir,
+		"--logdir", path.Join(g.LocalStateDir, "log"),
+		"--logfile", "glusterd2.log",
+	}
+	if externalEtcd {
+		args = append(args,
+			"--noembed",
+			// non-default port to avoid conflict with "real" etcd
+			// on system
+			// TODO: dynamic ports to avoid test conflicts?
+			"--etcdendpoints", "http://localhost:22379",
+		)
+	}
+	g.Cmd = exec.Command(path.Join(binDir, "glusterd2"), args...)
+
+	if err := g.Cmd.Start(); err != nil {
+		return nil, err
+	}
+
+	go func() {
+		g.Cmd.Wait()
+	}()
+
+	retries := 4
+	waitTime := 2000
+	for i := 0; i < retries; i++ {
+		// opposite of exponential backoff
+		time.Sleep(time.Duration(waitTime) * time.Millisecond)
+		if g.IsRestServerUp() {
+			break
+		}
+		waitTime = waitTime / 2
+	}
+
+	if !g.IsRestServerUp() {
+		return nil, errors.New("timeout: could not query rest server")
+	}
+
+	return &g, nil
+}
+
+// etcdProcess is used to manage etcd processes by the test suite
+// when glusterd2 is not running it's own embedded etcd.
+type etcdProcess struct {
+	testProcess
+	DataDir string
+	LogPath string
+}
+
+// Spawn starts a new etcd instance.
+func (ep *etcdProcess) Spawn() error {
+	args := []string{
+		"--name", "e2e-test-etcd",
+		"--data-dir", ep.DataDir,
+		"--listen-client-urls", "http://localhost:22379",
+		"--advertise-client-urls", "http://localhost:22379",
+		"--log-output", "stdout",
+	}
+	ep.Cmd = exec.Command("etcd", args...)
+
+	var (
+		logf *os.File
+		err  error
+	)
+	if ep.LogPath != "" {
+		logf, err = os.OpenFile(ep.LogPath,
+			os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+		if err != nil {
+			return err
+		}
+		ep.Cmd.Stdout = logf
+		defer logf.Close()
```

Comment:
Wouldn't this close the file ? Should this statement be moved to the goroutine spawned below which waits for the process to die ?

Reply:
:-) I'm glad you brought that up, I'll add a comment to clarify. In brief, once the process has been spawned the parent process no longer needs the fd and can close it. defer was used such that the file is closed on all paths.

Reply:
Indeed!